src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -0,0 +1,251 @@
+
+		private readonly BatchBlock<object> _eventQueue =
+			new BatchBlock<object>(20, new GroupingDataflowBlockOptions
+				{ BoundedCapacity = 1_000_000 });
+
+		private readonly HttpClient _httpClient;
+		private readonly IApmLogger _logger;
+
+		private readonly Service _service;
+
+		private CancellationTokenSource _batchBlockReceiveAsyncCts;
+
+		private readonly JsonSerializerSettings _settings;
+
+		private readonly SingleThreadTaskScheduler _singleThreadTaskScheduler = new SingleThreadTaskScheduler(CancellationToken.None);
+
+		public PayloadSenderV2(IApmLogger logger, IConfigurationReader configurationReader, Service service, HttpMessageHandler handler = null)
+		{
+			_settings = new JsonSerializerSettings { ContractResolver = new CamelCasePropertyNamesContractResolver(), Formatting = Formatting.None };
+			_service = service;
+			_logger = logger?.Scoped(nameof(PayloadSenderV2));
+
+			var serverUrlBase = configurationReader.ServerUrls.First();
+			var servicePoint = ServicePointManager.FindServicePoint(serverUrlBase);
+
+			servicePoint.ConnectionLeaseTimeout = DnsTimeout;
+			servicePoint.ConnectionLimit = 20;
+
+			_httpClient = new HttpClient(handler ?? new HttpClientHandler())
+			{
+				BaseAddress = serverUrlBase
+			};
+
+			if (configurationReader.SecretToken != null)
+			{
+				_httpClient.DefaultRequestHeaders.Authorization =
+					new AuthenticationHeaderValue("Bearer", configurationReader.SecretToken);
+			}
+			_creation = Task.Factory.StartNew(
+				() =>
+				{
+					try
+					{
+						_worker = DoWork();
+					}
+					catch (TaskCanceledException ex)
+					{
+						_logger.LogDebugException(ex);
+					}
+				}, CancellationToken.None, TaskCreationOptions.LongRunning, _singleThreadTaskScheduler);
+		}
+
+		private Task _worker;
+
+		public void QueueTransaction(ITransaction transaction)
+		{
+			_eventQueue.Post(transaction);
+			_eventQueue.TriggerBatch();
+		}
+
+		public void QueueSpan(ISpan span) => _eventQueue.Post(span);
+
+		public void QueueError(IError error) => _eventQueue.Post(error);
+
+		/// <summary>
+		/// Flushes all the events and ends the loop that processes and sends the events.
+		/// This can be called only once and after that the instance won't process anything.
+		/// </summary>
+		internal async Task FlushAndFinishAsync()
+		{
+			_logger.LogDebug("FlushAndFinish called - PayloadSenderV2 will become invalid");
+			await _creation;
+			_eventQueue.TriggerBatch();
+
+			_batchBlockReceiveAsyncCts.Cancel();
+
+			try
+			{
+				await _worker;
+			}
+			catch (TaskCanceledException)
+			{
+				_logger.LogDebug("worker task cancelled");
+			}
+			finally
+			{
+				_batchBlockReceiveAsyncCts.Dispose();
+			}
+
+			if (_eventQueue.TryReceiveAll(out var queueItems))
+				await ProcessQueueItems(queueItems.SelectMany(n => n).ToArray());
+
+			_eventQueue.Complete();
+		}
+
+		private async Task DoWork()
+		{
+			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
+			while (true)
```

Comment:
Small nitpicks : 

``` csharp
while (await _eventQueue.OutputAvailableAsync(_batchBlockReceiveAsyncCts.Token)) ?
```

It think its better to block on that then the `RecieveAsync` that follows: 

https://github.com/dotnet/corefx/blob/940ecee40b64ab93c92f87a7e8909771c013f1a5/src/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L1413


_batchBlockReceiveAsyncCts  can be readonly and assigned in the constructor.


Awesome PR 💯